### DISCUSSION TITLE
How to Utilize AD Stress Materials for the Homogenization system

### DISCUSSION THREAD

**Discussion by stonebeta** [idaholab/moose#27373] [Q&A Modules: Solid mechanics]
I'm interested in applying AD stress materials to homogenization problems. It appears that there isn't an AD version available for the Lagrangian system. Are there any straightforward methods currently available to achieve this? 

Alternatively, I've noticed methods like `ComputeLagrangianWrappedStress`. Perhaps converting AD stress materials to the Lagrangian system might be more convenient? However, I am unsure how to extract the pk1_jacobian from the AD stress material. Any guidance on this would be greatly appreciated.

**Comment by GiudGiud** [comment]
Hello

The new Lagrangian system does not use AD.
You could try to convert your AD material property to regular material properties with this Material
https://mooseframework.inl.gov/source/materials/MaterialADConverter.html

and lose the derivatives. You ll get an imperfect Jacobian but you could work with PJFNK for the nonlinear solver

Guillaume

**Reply by stonebeta** [reply]
Thank you for the response, but it seems that `MaterialADConverter` may not address this issue. Both in the `HomogenizationConstraint` and `TotalLagrangianStressDivergenceBase`, which is the parent class of `HomogenizedTotalLagrangianStressDivergence`, the material pk1_jacobian is required:
```C
HomogenizationConstraint::HomogenizationConstraint(const InputParameters & parameters)
  : ElementUserObject(parameters),
    _large_kinematics(getParam<bool>("large_kinematics")),
    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
    _F(getMaterialPropertyByName<RankTwoTensor>(_base_name + "deformation_gradient")),
    _pk1_stress(getMaterialPropertyByName<RankTwoTensor>(_base_name + "pk1_stress")),
    _pk1_jacobian(getMaterialPropertyByName<RankFourTensor>(_base_name + "pk1_jacobian"))
{
  ...
}
```
```C
template <class G>
TotalLagrangianStressDivergenceBase<G>::TotalLagrangianStressDivergenceBase(
    const InputParameters & parameters)
  : LagrangianStressDivergenceBase(parameters),
    _pk1(getMaterialPropertyByName<RankTwoTensor>(_base_name + "pk1_stress")),
    _dpk1(getMaterialPropertyByName<RankFourTensor>(_base_name + "pk1_jacobian"))
{
}
```
So, is there a method to explicitly extract pk1_jacobian as a material?

**Reply by GiudGiud** [reply]
@hugary1995

**Reply by hugary1995** [reply]
AD is not currently not supported with the homogenization constraints. I will add an error check in the action.
